Extracting commit contents

willdavidson05 opened the pull request:
<!-- _modified from [EmbeddedArtistry](https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/)_
_referenced with modifications from [pycytominer](https://github.com/cytomining/pycytominer/blob/master/.github/PULL_REQUEST_TEMPLATE.md)_ -->

# Description
This PR creates a `git_extracting.py` function that retrieves contents from a specific commit in a Git repository. The function can be tested using the test datasets created through the `add_data.py` function. Changes were also made to the test datasets and their contents, and the need for them to be zipped to maintain functionality is no longer necessary.

A test set was created to ensure that these test repositories are correctly deleted, to avoid a conflict with multiple repositories. 

I appreciate any comments or feedback!

Closes #48 
<!--
Thank you for your contribution to the Software Gardening Almanack!

Please _succinctly_ summarize your proposed change.
Namely, consider addressing the following questions:

- What motivated you to open this pull request?
- Were there any special adjustments you had to make to complete the work?
- Are there any issues which are related to this pull request (you may used a `#<digit>` to reference GitHub issues as links within this description)?

Also, if you haven't already, please use `pre-commit run --all-files` to help check your files using this project's pre-commit configuration.
-->

## What is the nature of your change?

- [x] Content additions or updates (adds or updates content)
- [ ] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).
- [ ] Breaking change (these changes would cause existing functionality to not work as expected).

# Checklist

Please ensure that all boxes are checked before indicating that this pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own contributions.
- [x] I have commented my content, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to related documentation (outside of book content).
- [x] My changes generate no new warnings.
- [x] New and existing tests pass locally with my changes.
- [ ] I have added tests that prove my additions are effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

d33bs commented on src/almanack/git_extracting.py:
```diff
@@ -0,0 +1,83 @@
+    return all_logs
+
+
+if __name__ == "__main__":
+    logs = main()
+    print(logs)
```
Consider removing this block to help this file become a module within the Python package instead of a script.

d33bs reviewed (commented):
Great job! I added a few comments with this review, please don't hesitate to let me know if you have any questions.

willdavidson05 commented:
Thanks for the review and feedback @d33bs !! I'll start working on these comments now.

willdavidson05 commented on src/almanack/git_extracting.py:
I didn't think of this, but you have a good point! I changed the name to git_parser, figured this gives a broader scope while still being intentional with git. But I welcome other ideas too, so let me know what you think!

willdavidson05 commented on tests/data/almanack/entropy/add_data.py:
```diff
@@ -13,42 +12,20 @@
-
-
-def commit_changes(directory: str, message: str):
+def commit_changes(directory: dir, message: str):
```
would it make sense to move the entire add_data file into the conftest as fixtures. After changing how my git function works and adding parts to the testing suite, I can see how this would work. Is this what you are suggesting and what are your thoughts on this?

willdavidson05 commented on src/almanack/git_extracting.py:
```diff
@@ -0,0 +1,83 @@
+        # Retrieve commit logs for each repository and store them in the dictionary
+        all_logs[repo_name] = get_commit_logs(repo_path)
+
+    return all_logs
```
Thanks for this! I moved the functionally of the repo paths over to the testing suite and passed it through a main in my git_parser file. Is this a good way of doing this? Or should I attempt this without a main?
```python
def main(repositories):
    all_logs = {}
    for repo_name, repo_path in repositories.items():
        # Retrieve commit logs for each repository and store them in the dictionary
        all_logs[repo_name] = get_commit_logs(repo_path)
    return all_logs
```

willdavidson05 commented on src/almanack/git_extracting.py:
```diff
@@ -0,0 +1,83 @@
+              Example: {'filename': 'file_content'}
+    """
+    contents = {}
+    repo = git.Repo(repository_path)
```
I'm a little confused with this. Should I pass the Repo object through as a argument in the function? Would you be able to give an example of how to do this?

willdavidson05 commented:
@d33bs Just a couple of other notes! I've added `gitpython` to the dependencies of poetry to ensure that the testing suite runs smoothly. Also, I've successfully implemented the creation, reading, and deletion of my test repositories (high and low entropy) using `conftest.py`. I'm curious if you think `conftest.py` is the appropriate place for this functionality, or if it should be moved to its own separate test file? Thanks again for your review and feedback! I've requested another review with the necessary changes. No rush, but please let me know your thoughts!

d33bs commented on tests/conftest.py:
```diff
@@ -74,3 +74,39 @@ def build_jupyter_book(
+    yield repositories
+
+
+def test_main(repository_paths):
```
Consider moving this test to a specific module which focuses on tests for `git_parser` (or another form of organization, such as within `test_almanack`). Generally the pattern is to keep fixtures and elements directly related to fixtures within `conftest.py` and to keep tests in separate modules which are named after the capabilities they test for the package. The goal is to ensure the test is run when one issues `poetry run pytest`.

d33bs commented on tests/conftest.py:
```diff
@@ -74,3 +74,39 @@ def build_jupyter_book(
+    assert "low_entropy" in all_logs
+
+    # Delete the high_entropy and low_entropy directories to sustain repository structure
+    for repo_path in repository_paths.values():
+        shutil.rmtree(repo_path)
```
This looks like it might be good to include after the yield in the fixture instead of the test. Generally we want the tests to be isolated and specific to testing capability instead of scaffolding / de-scaffolding for the tests themselves (where we can help it).

d33bs commented on tests/data/almanack/entropy/add_data.py:
```diff
@@ -13,36 +12,14 @@
 import pathlib
 import subprocess
-import zipfile
-from pathlib import Path
```
It looks like we might need this import below - consider keeping it if so.

d33bs commented on src/almanack/git_parser.py:
```diff
@@ -0,0 +1,58 @@
+    return contents
+
+
+def main(repositories):
```
Consider adding a docstring for this function.

d33bs commented on src/almanack/git_parser.py:
```diff
@@ -0,0 +1,58 @@
+import git
+
+
+def get_commit_logs(repository_path: str):
```
Consider adding a test for this function within the test suite.

d33bs commented on src/almanack/git_parser.py:
```diff
@@ -0,0 +1,58 @@
+    return logs
+
+
+def get_commit_contents(repository_path: str, commit_id: str):
```
Consider adding a test for this function within the test suite.

d33bs reviewed (commented):
Nice job - thanks for addressing those comments! I've added few more comments with this review.

Generally as we add functions we want to ensure they have "coverage" through additional tests. While it's nearly impossible and might be unreasonable to achieve 100% test coverage, a good rule of thumb is to seek ~90%. As you add tests which implement the functions you'll notice the coverage percentage increase.

To get a quick visual on this, you could try out the following:
```sh
# add coverage to the env
poetry add coverage
# run pytest with via coverage
poetry run coverage run -m pytest
# export the coverage report in html
poetry run coverage html
# open the coverage display within a browser
open htmlcov/index.html
# click on the files to see what's covered or not
```

willdavidson05 commented:
Thanks for the review @d33bs !! I've implemented almost all of your feedback! Would love to hear what you think about the test cases I have added. These tests were what I could come up with, but would like to hear if you have any ideas! I will note, that the coverage was 100% for these tests. I will go ahead and request a re-review, but no rush. Thank you!

d33bs commented on pyproject.toml:
```diff
@@ -15,6 +15,8 @@ exclude = ["src/book/_build"]
 python = ">=3.9,<=3.12"
 pyyaml = "^6.0.1"
+gitpython = "^3.1.43"
+coverage = "^7.5.3"
```
Consider moving `coverage` to a `[tool.poetry.group.dev.dependencies]` group dependency (it would only be needed for development purposes, and not necessary for the audience using the package).

d33bs commented on tests/conftest.py:
```diff
@@ -74,3 +74,30 @@ def build_jupyter_book(
+    Fixture to call add_Data.py, create the repositores, then delete them
+    """
+    # Compute the base path relative to this location
+    base_path = (
```
Consider the use of the built-in [`tmp_path`](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html) fixture which can provide a temporary directory for use with tests. The nice part about these is that they are automatically removed once processing completes. Using this might also involve adding an additional parameter for directory specification within `add_data()`.

d33bs commented on tests/conftest.py:
```diff
@@ -74,3 +74,30 @@ def build_jupyter_book(
+
+    yield repositories
+    # Delete the high_entropy and low_entropy directories to sustain repository structure
+    shutil.rmtree(base_path / "high_entropy", ignore_errors=True)
+    shutil.rmtree(base_path / "low_entropy", ignore_errors=True)
```
If you opt for the `tmp_path` above, consider removing these lines.

d33bs commented on tests/data/almanack/entropy/add_data.py:
```diff
@@ -51,36 +17,42 @@ def commit_changes(directory: str, message: str):
+    repo.index.commit(message)
+
 
+def add_data():
```
Consider using a different name for this function or the module (filename) to help keep these distinct.

d33bs commented on tests/test_git_parser.py:
🙌  great work with this!

d33bs reviewed (approved):
Great work! I left a few minor comments and overall thought this LGTM. Please feel free to merge when happy with things.
